### ISSUE TITLE
Music loop has a small gap between songs

### ISSUE THREAD

**Issue by ericoporto** [adventuregamestudio/ags#501]
There is a small gap when a music loops. 

![image](https://user-images.githubusercontent.com/2244442/44963242-5aea5400-aeff-11e8-848f-c3bcf1a27571.png)

Something like 0.032 s . I know it's small but it can be listened.

**Comment by ericoporto** [comment]
I will make a minimal game, playing a constant tone and attach here. It happens on both Windows and Linux for me.

**Comment by ivan-mogilko** [comment]
Oh, correct, it's possible to use constant tone to test this, I guess no need to attach a game then.

**Comment by ericoporto** [comment]
Adding a continuous tone.
[SinWaveTest_1.wav.zip](https://github.com/adventuregamestudio/ags/files/2346777/SinWaveTest_1.wav.zip)
![image](https://user-images.githubusercontent.com/2244442/45004402-926e0480-afc2-11e8-9aba-347a833371d2.png)

**Comment by morganwillcock** [comment]
Just to check, what format was this constant tone saved in, if it also had the same issue?

**Comment by ivan-mogilko** [comment]
Sorry, I forgot to confirm that I could hear the "gap", although it's indeed really small.

I think it would be interesting to test if it's length depends on fps.

**Comment by morganwillcock** [comment]
I don't think this file is a valid test; I hear a click when looping (in an audio editor, so nothing to do with AGS), but that is because the start of the file doesn't match the end.

![loop](https://user-images.githubusercontent.com/14958747/45932405-91057b80-bf73-11e8-9f06-01a6de33b8a6.png)

**Comment by ivan-mogilko** [comment]
I absolutely don't know if related (and we don't have a confirmed clip example yet?), but just in case here's an old commit I found when investigating separate issue: 688e7d88192f45e64cf9593468cdc046f1fa7a2c

**Comment by morganwillcock** [comment]
Just a thought:

Given the size of the gap here (0.032) and the size of the gap which someone else has reported (~ 20ms), this is very similar to the gap for 1 engine step at the default framerate (1/40 = 0.025). I would imagine that WAV playback or anything using the threaded audio playback option may get close to a seamless loop, but it would be down to luck on whether the engine step happened to be in-sync with the end of the audio track. For the normal OGG and MP3 playback, I guess this is buffering chunks of audio large enough to cover the gap between engine steps, so the final chunk is always likely to have a gap.

**Comment by ivan-mogilko** [comment]
These were my thoughts too, and I don't know if there may be any quick fix for this. Engine would probably need to have a different kind of audio queue, which processes all queued/looped clips seamlessly without waiting for the next step to change the active clip.

**Comment by tzachshabtay** [comment]
You should be able to verify that's this is indeed the real cause, then, by changing the framerate and seeing if it changes the gap size.

**Comment by ivan-mogilko** [comment]
Actually, looping clips work differently, so maybe these are really two separate cases...

**Comment by jjattar** [comment]
I'm the guy who [reported the 20ms gap](http://www.adventuregamestudio.co.uk/forums/index.php?topic=56598.0) (my problem is with the PlayQueued command, loops play fine). After reading all this, I did some testing that confirms something I've suspected: the length of the gap changes each time. I've recorded the same part of the game 4 times, gap was present in all of them, and the length was different each and every time (the longest is around 18ms). This should confirm the engine step thing, right?

**Comment by sonneveld** [comment]
@ericoporto Old issue, but do you remember what format the audio clip was in? Does anyone know off the top of their head if cross fading works between queued tracks?

I think the correct fix would be to process queued audio and cross fading within the audio update thread. If a clip ends while filling the buffer, we could continue with the next clip.

**Comment by ericoporto** [comment]
I don't remember the first audio what it was,  but my second test audio was added above.

**Comment by ivan-mogilko** [comment]
@ericoporto will you be able to test with the 3.6.0? I think our audio system now works similar to how @sonneveld mentioned.

**Comment by ericoporto** [comment]
I can test this on the weekend, please assign the issue for me just so I don't forget.

**Comment by ericoporto** [comment]
[sineTest.zip](https://github.com/adventuregamestudio/ags/files/7796886/sineTest.zip)

It took a long time, ha, but anyway, there is a little gap that is audible in this test (around each 6 seconds). But I don't really know if it was supposed to BE audible here, because I did record it with Audacity and it's impossible to tell where it is - I spent a long time looking and couldn't figure out.

**Comment by ericoporto** [comment]
[AudioTests.zip](https://github.com/adventuregamestudio/ags/files/7964829/AudioTests.zip)

Grundislav was careful to make a nice report on audio gap, I attached it above. I also have been thinking it could possibly be an audio pop. Popping sounds could originated for different reasons than just a gap, I think we previously had a gap but now it could be a different similar thing.

**Comment by ivan-mogilko** [comment]
In regards to this:

> [sineTest.zip](https://github.com/adventuregamestudio/ags/files/7796886/sineTest.zip)

It took a long time, ha, but anyway, there is a little gap that is audible in this test (around each 6 seconds). But I don't really know if it was supposed to BE audible here, because I did record it with Audacity and it's impossible to tell where it is - I spent a long time looking and couldn't figure out.

I exported the sine wav from the audio.vox, and played looped in Audiacity, and it has same "pop" where it ends previous instance and starts a new one. I even pasted several copies of the same soundwave to make a longer wave, and the result had a "pop" each 6 seconds, where the end connects to the new beginning. So I guess this is a looping imprefection, similar to the one noted by morganw earlier in this ticket.

**Comment by ivan-mogilko** [comment]
In regards to this.

> [AudioTests.zip](https://github.com/adventuregamestudio/ags/files/7964829/AudioTests.zip)

> Grundislav was careful to make a nice report on audio gap, I attached it above

Unfortunately, this test does not look good. There's a readme that tells what the folders are, but is no explanation what to pay attention to in these test games, nor the recorded audio from games. In test games themselves there's no indication of current song position, thus it's not possible to know when to expect going on a second loop. I cannot even tell whether I am supposed to look for a loop, or just a random gap / popping sound in a random place. I wish I had a chance to check this out and talk to a game author back when this was prepared.

**Comment by ivan-mogilko** [comment]
Alright, so, after speaking with Grundislav, there are two situations here:

1. Scheduling next audio with PlayQueue. It still has a gap between two songs, although that's not what you'd expect from the PlayQueue function. I believe that may be because the queue is updated only once per game frame, which is way too slow to react to the audio thread. If that's the case, this means the whole audio queue mechanism will have to be *rewritten*.
2. Scheduling next audio by testing for the previous clip's state in rep-exec-always. This cannot work reliably for similar reasons, but also because of this change in audio handling logic:
> Engine now ensures that in threaded audio mode the AudioChannel's state only changes once the
   game is updated, and not while the game script is running. This prevents situations when
   the clip could begin playing before all properties are set in script, or when AudioChannel's
   properties' values (such as Position) could change while running a game script.
In theory p2 could be improved by also updating channel states in post-script-cleanup, that is when the current callback ends. But there still won't be a guarantee of a seamless transition. In general such approach (controlling sound order in script in real-time) is not correct if one wants a full precision, as running script itself takes time, and it's not known how much time have passed between audio have stopped and script callback was run.


I think I will open a new ticket for PlayQueue, because it's not related to sound looping per se. In regards to the second problem, I can make an experimental build and send to Grundislav to let him see if that improves anything.

**Comment by ivan-mogilko** [comment]
Opened #1687 for the sound queue (PlayQueued).

I think this may be closed, as the original issue (gap between sound loops) seem fixed?

**Comment by ivan-mogilko** [comment]
> Just to clarify, the mechanism of a music that is set to repeat/loop is the same of PlayQueued? (Like, per frame, detect it has finished and ask it to play again)

No, they are currently not related at all.

**Comment by ericoporto** [comment]
Thanks for the clarification. Closing this issue, the PlayQueued stuff can be solved in it's own then. :)

**Comment by ericoporto** [comment]
Someone reported this again on the forums...: https://www.adventuregamestudio.co.uk/forums/index.php?topic=61098.0

**Comment by morganwillcock** [comment]
> Someone reported this again on the forums...: https://www.adventuregamestudio.co.uk/forums/index.php?topic=61098.0

I'm pretty sure this is a false alarm. I've tested it and converting their audio file from mp3 to ogg fixes the problem.
